### PR TITLE
Cics SOAPFAULT improvements

### PR DESCRIPTION
**Pull request by mm-broadcom**
## How Has This Been Tested?

Added both valid and invalid tests to this and ran them against the error checking.

## Checklist:
- [x] Each of my commits contains one meaningful change
- [x] I have performed rebase of my branch on top of the development
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by mm-broadcom** [comment]
@ilidio-lopes Could you review this?

File: server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCICSSoapfaultStatement.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                    + "       DATA DIVISION.\n"
+                    + "       WORKING-STORAGE SECTION.\n"
+                    + "       PROCEDURE DIVISION.\n"
+                    + "            EXEC CICS \n"
+                    + "            END-EXEC.";
+
+    // Test Strings
+    private static final String SOAPFAULT_DELETE_VALID = "SOAPFAULT DELETE";
+
+    private static final String SOAPFAULT_ADD_1 = "SOAPFAULT ADD FAULTSTRING(123) FAULTSTRLEN(123) NATLANG(123) FROMCCSID(123)";
```
Comment (chacebot) [inline server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCICSSoapfaultStatement.java]: Variables

File: server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4 [hunk 2]
```diff
@@ -670,17 +670,10 @@ cics_signon: SIGNON (USERID cics_data_value | ESMREASON cics_data_area | ESMRESP
...
-cics_soapfault_create: CREATE (FAULTCODE cics_cvda | CLIENT | SERVER | SENDER | RECEIVER | cics_soapfault_faultcodestr |
-                       cics_soapfault_faultstring | cics_soapfault_role | cics_soapfault_faultactor |
-                       cics_soapfault_detail | FROMCCSID cics_data_value | cics_handle_response)+;
-cics_soapfault_faultcodestr: (FAULTCODESTR cics_data_value | FAULTCODELEN cics_data_value)+;
-cics_soapfault_role: (ROLE cics_data_value | ROLELENGTH cics_data_value)+;
-cics_soapfault_faultactor: (FAULTACTOR cics_data_value | FAULTACTLEN cics_data_value)+;
-cics_soapfault_detail: (DETAIL cics_data_value | DETAILLENGTH cics_data_value)+;
+cics_soapfault: SOAPFAULT (cics_soapfault_add | cics_soapfault_create | DELETE cics_handle_response*);
+cics_soapfault_add: ADD ((FAULTSTRING | FAULTSTRLEN |NATLANG | SUBCODESTR | SUBCODELEN | FROMCCSID) cics_data_value | cics_handle_response)+;
+cics_soapfault_create: CREATE ((CLIENT | SERVER | SENDER | RECEIVER | NATLANG) | (FAULTCODESTR | FAULTCODELEN | FAULTSTRING | FAULTSTRLEN | NATLANG | ROLE | ROLELENGTH | FAULTACTOR | FAULTACTLEN | DETAIL | DETAILLENGTH | FROMCCSID) cics_data_value
```
Comment (chacebot) [inline server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4]: The first NATLANG should not be parameterless and should accept a string literal parameter that is checked for 'en' in the parser validation.

Comment (mm-broadcom) [reply server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4]: We cannot enforce the use of 'en', that is merely the default language option.

Comment (mm-broadcom) [reply server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/cics/CICSParser.g4]: Fixed the grammar in latest commit.

File: server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCICSSoapfaultStatement.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    // Test Strings
+    private static final String SOAPFAULT_DELETE_VALID = "SOAPFAULT DELETE";
+
+    private static final String SOAPFAULT_ADD_1 = "SOAPFAULT ADD FAULTSTRING({$varOne}) FAULTSTRLEN({$varOne}) NATLANG({$varOne}) FROMCCSID({$varOne})";
+    private static final String SOAPFAULT_ADD_2 = "SOAPFAULT ADD FAULTSTRING({$varOne})";
+
+    private static final String SOAPFAULT_CREATE_1 = "SOAPFAULT CREATE FAULTCODE({$varOne}) FAULTSTRING({$varOne}) FAULTSTRLEN({$varOne}) NATLANG({$varOne}) ROLE({$varOne}) ROLELENGTH({$varOne}) FAULTACTOR({$varOne}) FAULTACTLEN({$varOne}) DETAIL({$varOne}) DETAILLENGTH({$varOne}) FROMCCSID({$varOne})";
+    private static final String SOAPFAULT_CREATE_2 = "SOAPFAULT CREATE CLIENT FAULTSTRING({$varOne})";
+
+    // Invalid Tests
```
Comment (slavek-kucera) [inline server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCICSSoapfaultStatement.java:45]: Please add a test for duplicate checking of the `DELETE` subcommand.

File: server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSoapfaultOptionsCheckUtility.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                break;
+            default:
+                break;
+        }
+
+        checkDuplicates(ctx);
+    }
+
+    private void checkCreate(CICSParser.Cics_soapfault_createContext ctx) {
+        checkMutuallyExclusiveOptions("FAULTCODESTR, CLIENT, SERVER, SENDER or RECEIVER", ctx.FAULTCODESTR(), ctx.CLIENT(), ctx.SERVER(), ctx.SENDER(), ctx.RECEIVER());
```
Comment (slavek-kucera) [inline server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSoapfaultOptionsCheckUtility.java]: exactly one of ... and FAULTCODE?

File: server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSoapfaultOptionsCheckUtility.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                checkAdd((CICSParser.Cics_soapfault_addContext) ctx);
+                break;
+            default:
+                break;
+        }
+
+        checkDuplicates(ctx);
+    }
+
+    private void checkCreate(CICSParser.Cics_soapfault_createContext ctx) {
```
Comment (slavek-kucera) [inline server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSoapfaultOptionsCheckUtility.java:87]: FAULTSTRING is required.

File: server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSoapfaultOptionsCheckUtility.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        checkPrerequisiteIsMet(ctx.FAULTACTOR(), ctx.FAULTACTLEN(), ctx, "FAULTACTLEN without FAULTACTOR");
+
+        checkPrerequisiteIsMet(ctx.DETAIL(), ctx.DETAILLENGTH(), ctx, "DETAILLENGTH without DETAIL");
+
+        checkPrerequisiteIsMet(ctx.FAULTCODE(), ctx.FAULTCODELEN(), ctx, "FAULTCODELEN without FAULTCODE");
+
+        checkPrerequisiteIsMet(ctx.FAULTSTRING(), ctx.FAULTSTRLEN(), ctx, "FAULTSTRLEN without FAULTSTRING");
+    }
+
+    private void checkAdd(CICSParser.Cics_soapfault_addContext ctx) {
```
Comment (slavek-kucera) [inline server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSoapfaultOptionsCheckUtility.java:105]: exactly one of FAULTSTRING or SUBCODESTR

File: server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCICSSoapfaultStatement.java (hunk 3 above)
Comment (mm-broadcom) [reply server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCICSSoapfaultStatement.java:45]: Added in latest commit.

File: server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSoapfaultOptionsCheckUtility.java (hunk 4 above)
Comment (mm-broadcom) [reply server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSoapfaultOptionsCheckUtility.java]: Added in latest commit.

File: server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSoapfaultOptionsCheckUtility.java (hunk 5 above)
Comment (mm-broadcom) [reply server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSoapfaultOptionsCheckUtility.java:87]: Added in latest commit.

File: server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSoapfaultOptionsCheckUtility.java (hunk 6 above)
Comment (mm-broadcom) [reply server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSoapfaultOptionsCheckUtility.java:105]: Added in latest commit.

File: server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCICSSoapfaultStatement.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    private static final String SOAPFAULT_DELETE_INVALID = "SOAPFAULT DELETE {CLIENT|errorOne}";
+    private static final String SOAPFAULT_DELETE_INVALID_2 = "SOAPFAULT DELETE DELETE {NORESP|errorOne}";
```
Comment (slavek-kucera) [inline server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/TestCICSSoapfaultStatement.java]: `NORESP` is not allowed, however...
```
 LINE           SOURCE LISTING                                                                                      
                                                                                                                    
00001                     EXEC CICS                                                00070000                         
00002                     SOAPFAULT                                                00070101                         
00003                     DELETE                                                   00070201                         
00004                     DELETE                                                   00070301                         
00005                     NOHANDLE                                                 00070401                         
00006                     END-EXEC                                                 00071000                         
00007                                                                              00079400                         
CICS 7.3.0 COMMAND LANGUAGE TRANSLATOR                                            TIME 07.16 DATE 16 JAN 25   PAGE 2
TRANSLATOR DIAGNOSTIC MESSAGES OF SEVERITY W AND ABOVE.                                                             
MESSAGE NUMBER  SEVERITY  LINE NUMBER     MESSAGE                                                                   
DFH7036I           W        00001         PROGRAM DOES NOT BEGIN WITH AN IDENTIFICATION DIVISION.                   
DFH7057I           W        00001         DUPLICATE SPECIFICATION FOR DELETE OPTION IS IGNORED.                     
END OF TRANSLATOR DIAGNOSTIC MESSAGES.                                                                              
```

File: server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSoapfaultOptionsCheckUtility.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                    put(CICSLexer.ROLELENGTH, ErrorSeverity.ERROR);
+                    put(CICSLexer.FAULTACTOR, ErrorSeverity.ERROR);
+                    put(CICSLexer.FAULTACTLEN, ErrorSeverity.ERROR);
+                    put(CICSLexer.DETAIL, ErrorSeverity.ERROR);
+                    put(CICSLexer.DETAILLENGTH, ErrorSeverity.ERROR);
+                    put(CICSLexer.FROMCCSID, ErrorSeverity.ERROR);
+
+                    put(CICSLexer.DELETE, ErrorSeverity.ERROR);
+                    put(CICSLexer.CREATE, ErrorSeverity.ERROR);
+                    put(CICSLexer.ADD, ErrorSeverity.ERROR);
```
Comment (slavek-kucera) [inline server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSSoapfaultOptionsCheckUtility.java]: Warnings, as can be seen in the snippet from the previous comment.

**Review by slavek-kucera — DISMISSED** [review]

**Review by slavek-kucera — APPROVED** [review]
